Remove deprecated compiler Makefiles

kinke opened the pull request:
To check if CI still needs them somewhere.

kinke committed:
Remove deprecated compiler Makefiles

kinke committed:
Remove deprecated compiler/test/Makefile

kinke committed:
Top-level posix.mak: Include druntime in `test` and `clean` rules

This probably pre-dates the druntime merge into the dmd repo.

PetarKirov commented:
I'm in favor of the removal, but please add a changelog entry as this a breaking change for package maintainers.

kinke commented:
The `DAutoTest` CI job would need to be adapted; no idea where.

Breaking changes OTOH:
* No more deprecated compiler *build* Makefiles (`compiler/src/{posix,win32,win64}.mak`) - superseded long ago by `build.d`.
* No more deprecated compiler *test* Makefile (`compiler/test/Makefile`) - superseded long ago by `run.d`.
* Top-level Makefiles:
  * `posix.mak` should be pretty compatible, except for the druntime addition to `test` and `clear` rules (could be extracted to its own PR though). I'm not sure about the Make variables in the cmdline though, e.g., `make ENABLE_RELEASE=1`; specifying those as env vars should work (used by `build.d` then).
  * The Windows `win{32,64}.mak` have been removed; they only supported a single target, building the compiler alone. So I guess they are unused.
* Legacy Makefiles in `src/`:
  * The Posix `src/posix.mak` forwards to the top-level `posix.mak` now and should be pretty compatible (though e.g. `all` includes druntime now, not just the compiler itself).
  * The Windows `src/win{32,64}.mak` files were removed. Hopefully no tools depend on these.

kinke commented:
> I'm not sure about the Make variables in the cmdline though

Should be good, according to https://www.gnu.org/software/make/manual/html_node/Environment.html:
> By default, only variables that came from the make’s environment or set on its command line are placed into the environment of the commands.

kinke committed:
Convert posix.mak to generic Makefile

Usable on Windows too, with a GNU make.

kinke committed:
Restore posix.mak, forwarding to generic Makefile

ibuclaw commented:
@kinke looks like an environment variable is being ignored in @CyberShadow 's dautotest pipeline.

```
'HOST_DC=/home/dtest/DAutoTest/work/dl/dmd-2.079.0/dmd2/linux/bin64/dmd' '-j' '8'
make -C .. all
make[1]: Entering directory '/tmpfs/dautotest-work/repo/dmd'
dmd -ofgenerated/build -g compiler/src/build.d
make[1]: dmd: No such file or directory
make[1]: *** [Makefile:28: generated/build] Error 127
```

ibuclaw commented on compiler/src/posix.mak:
```diff
@@ -1,217 +0,0 @@
-  HOST_CXX=$(HOST_CC)
-endif
-
-HOST_DC?=
-ifneq (,$(HOST_DC))
-  $(warning ========== Use HOST_DMD instead of HOST_DC ========== )
-  HOST_DMD=$(HOST_DC)
-endif
```
Possibly this is what's missing in the makefile.

kinke replied:
Oh right, thx for looking! I thought `DAutoTest` was using the `compiler/src/posix.mak` (removed without proxy), but no, it's the `src/posix.mak` one.

kinke replied:
Oh well, it does use the removed one sometime later apparently:
```
make --directory=../dmd/compiler/src -f posix.mak AUTO_BOOTSTRAP=1 BUILD_JOBS=
make[1]: Entering directory '/tmpfs/dautotest-work/repo/dmd/compiler/src'
make[1]: posix.mak: No such file or directory
```

ibuclaw replied:
I suspect that is the dlang.org repo though.

https://github.com/dlang/dlang.org/blob/master/posix.mak#L596-L600

kinke committed:
Makefile: Improve search for D host compiler

ibuclaw commented:
👍 🙈

kinke committed:
Add changelog entry

kinke committed:
Top-level Makefile: Extract {dmd,druntime}[-test] targets

For convenience, and to make the compiler build triggered from
druntime/posix.mak more robust.

kinke commented on druntime/posix.mak:
```diff
@@ -350,8 +350,8 @@ $(IMPDIR)/%.h : src/%.h
 
-$(DMD):
-	$(MAKE) -C $(DMD_DIR)/src -f posix.mak BUILD=$(BUILD) OS=$(OS) MODEL=$(MODEL)
+$(DMD_DIR)/../generated/$(OS)/$(BUILD)/$(MODEL)/dmd:
```
This is a subtle but important change. E.g., for CI we specify a `DMD` var explicitly in the druntime make command-line (in some cases at least). Don't pretend that we are able to generate *that* DMD (in case it's non-existent) with the following recipe - the output path is fixed.

kinke commented on druntime/posix.mak:
```diff
@@ -350,8 +350,8 @@ $(IMPDIR)/%.h : src/%.h
-$(DMD):
-	$(MAKE) -C $(DMD_DIR)/src -f posix.mak BUILD=$(BUILD) OS=$(OS) MODEL=$(MODEL)
+$(DMD_DIR)/../generated/$(OS)/$(BUILD)/$(MODEL)/dmd:
+	$(MAKE) -C $(DMD_DIR)/.. dmd BUILD=$(BUILD) OS=$(OS) MODEL=$(MODEL) DMD=""
```
In case `DMD` is set in the druntime `make` cmdline (semantics: use that DMD to compile druntime), it's automatically added as env variable, and affects the host-compiler lookup. If we get here to this recipe, `DMD` is the default one, *and* non-existent, so we obviously can't use that one as host compiler. So clear `DMD` explicitly here.

kinke commented:
Prerequisite: https://github.com/dlang/installer/pull/563

thewilsonator commented:
This is green, is it good to go?

thewilsonator reviewed (approved):
LGTM
